### PR TITLE
Deployability — Hub push, inference container, droplet deploy (#302)

### LINKED ISSUE #302: Deployability — make inference + retraining portable to any machine
## Context

The project is graded as a software product, not just a research repo. A research repo that runs only on the author's machine fails the product bar regardless of how strong the methodology is. Today inference depends on local `/data/` mounts, local model checkpoints, and a local Python environment. Retraining additionally depends on local training packages and embedding caches. Both must be portable to a fresh machine before submission.

This is a finalization-roadmap lane that was missing from §16 at the time it was filed. See `fed-pulse.wiki/16_Finalization_Roadmap.md` and umbrella #299.

## Acceptance

A machine that has only Docker + an HF token + (for the reproducibility lane) the public repo can:

- Pull the inference container, start it, hit `/analyze` from a browser, and receive a valid response
- Reach the deployed inference surface at a public URL the examiner can hit
- Run `make reproduce-all` (or equivalent documented command) that fetches the canonical training package + embedding caches from HF and runs a training smoke against them

That is the "deployable on someone else's machine" test, and it is the closing bar for this issue.

### Stage 1 — Inference artefacts on the Hub

- Push canonical encoder backbones to HF Hub (FinBERT-FedAdjacent variants + cross-bank pretrained checkpoint once it lands) as private model repos
- Push the canonical multi-head forecaster checkpoint to HF Hub once #292 closes
- Add `hf://` URI support to the registry resolver alongside local paths

### Stage 2 — Training package portability

- Push canonical training package (`tp_v3_macro_aug_2026_05_25` or successor) to HF Datasets
- Update `build_training_package.py` and the package loader to accept either a local path or an HF Dataset reference
- Same `revision=` pinning convention as the upstream `gtfintechlab` inputs

### Stage 3 — Embedding cache portability

- Push canonical embedding caches (~1-2 GB total) to HF Hub
- Lazy-fetch on first use with a local cache as fall-through
- Verify hash on download

### Stage 4 — Inference container + cloud deploy

- `Dockerfile` that loads every artefact via `from_pretrained` (or equivalent HF Hub fetch) at boot; no `/data/` mount required
- Deploy the backend container to Fly.io / Railway / Render with the HF token in env
- Deploy the Next.js frontend to Vercel
- Public URL documented in `README.md` and the wiki

### Stage 5 — Reproducibility script + docs

- `make reproduce-all` (or `scripts/reproduce.sh`) that fetches the training package + caches from HF and runs a training smoke
- End-to-end documentation in `docs/reproduce.md`: clone, set HF token, run one command, training smoke completes

## Out of scope

- Push-button hyperparameter sweep on cloud GPUs (research workflow, not product)
- Multi-region deploy, autoscaling, observability stack
- Anything that touches the existing model architecture or training code beyond the path-resolution layer

## Dependencies

- Stage 1 partially depends on #292 closing so the canonical multi-head forecaster checkpoint exists. Encoder backbones can be pushed earlier.
- Stages 2–5 can run in parallel once Stage 1 lands.
- Cross-bank pretraining (parallel session) feeds the encoder push in Stage 1.

## Notes on PaaS choices

- Backend container expected size < 1 GB (FinBERT ~ 440 MB + heads ~ 50 MB + Python deps). Comfortable on free tiers.
- Cold-start latency 5–15 s on free PaaS; warm-up step documented for the demo run.
- Frontend on Vercel is free-tier and deploys from a `vercel.json` push.

## Reference

- Umbrella: #299
- Wiki: §16 Finalization Roadmap

### PR DESCRIPTION
**Pull request by yusufizzetmurat**
Closes #302.

- `hf://` URI resolver in `app.models.registry`; `artefacts:` block lists canonical Hub mirrors
- Training-package loader + embedding-cache `ensure_local` fall through to HF Hub on local miss
- `Dockerfile.prod` + `compose.prod.yml` + `deploy/Caddyfile` + s6-supervised entrypoint with eager-pull
- `.github/workflows/deploy.yml` (push to `main`) with health-probe rollback; `promote.yml` for dev->main
- Frontend prod build appends `/api` for Caddy strip; standalone output enabled
- `make reproduce-all` + `scripts/reproduce_all.py` + `docs/deploy.md` + `docs/reproduce.md`
- `scripts/push_artefacts_to_hub.py` ships with `--dry-run`; operator pushes once post-merge

Operator follow-ups: `python scripts/push_artefacts_to_hub.py --all`, pin returned shas in `registry.yaml`, set GH secrets (`DROPLET_SSH_KEY`, `HF_TOKEN`, `FRED_API_KEY`), provision per `docs/deploy.md`.

## Test plan
- `docker compose -p fed-pulse-bundlee run --rm backend pytest tests/unit/test_hf_registry_resolver.py tests/unit/test_embedding_cache_lazy_fetch.py tests/unit/test_reproduce_all_smoke.py -q`
- `docker compose -p fed-pulse-bundlee run --rm backend ruff check app/models/registry.py app/data/embedding_cache.py app/training/loaders.py`
- `cd frontend && npm run build`

### CODE REVIEW

**Comment by yusufizzetmurat** [comment]
Five-angle review fix-up landed in `884bda3`.

**Container security**
1. Non-root `fedpulse` UID 10001 via `s6-setuidgid` in service `run` scripts; bind-mount writes no longer land as root on the host.
2. Hash-pinned `backend/requirements.lock` (3705 lines, 30+ deps); `Dockerfile.prod` installs with `--require-hashes` and copies source AFTER the dep layer.
4. s6-overlay 3.1.6.2 download sha256-verified via `sha256sum -c`; version pinned via ARG.

**Build / deploy correctness**
3. Every `actions/*` reference pinned to a commit sha (v4.2.2 for checkout, v5.6.0 for setup-python).
5. `NEXT_PUBLIC_API_URL` plumbed as build ARG + ENV in `builder-frontend`; `compose.prod.yml build.args` bakes the production origin into the browser bundle.
6. `hf-cache` named volume mounted on both backend + frontend; `HF_HOME` env points there.
7. `PRIOR_SHA` captured via `git rev-parse HEAD` before `git fetch`; `last-deploy.sha` written only after the health probe succeeds.
8. Rollback re-runs the smoke probe; original failure exits 1, rollback-also-failed exits 2 with a clear log.
10. HF login failure is fatal — entrypoint.sh exits non-zero and uvicorn never starts on a broken token.
11. Source copy lands after `pip install` (fixed alongside #2).
17. New `wait-for-ci` job polls `ci.yml` on the same sha (10-min timeout) before any droplet ssh.

**Application code**
9. `push_artefacts_to_hub.py` stages model card in a tempdir; canonical data directories on disk are never mutated.
12. `upload_folder` ignore_patterns excludes `.env`, `.git`, credentials, `*.pem`, `*.key`, ipynb checkpoints, pycache; dry-run prints the per-file plan.
13/18. `parse_hf_uri` rejects path traversal (`../..`), trailing slashes, multi-colon URIs, empty revisions, and illegal revision characters via explicit regex; six new unit tests.
14. `reproduce_all` drops the `if dst.exists(): continue` skip — `copytree(dirs_exist_ok=True)` overwrites on revision bumps.
15. `ensure_local` wraps HF errors into typed `EmbeddingCacheUnavailable`; wired into `_read_chunk_embedding_lookup` so the production loader logs a fixable error instead of silent zero-embeddings.
16. Unused `FED_PULSE_REGISTRY_PREFER_HF` removed from `.env.example`, `Dockerfile.prod`, `compose.prod.yml`.
19. Caddyfile adds `Content-Security-Policy` + global `servers { trusted_proxies static private_ranges }`.
20. Frontend service threads `env_file` for config symmetry; `FED_PULSE_SKIP_EAGER_PULL=1` keeps it from hitting HF.
21. `docs/deploy.md` reorders DNS A-record + `dig` propagation check BEFORE `docker compose up`.
22. New `.github/workflows/hf-push.yml` workflow_dispatch with dry-run + repos inputs; uses `HF_TOKEN_WRITE` from GH secrets only.
23. Post-upload `model_info` / `dataset_info` call extracts the sha; missing sha raises `RuntimeError` so operator never pastes `?` into registry.yaml.
24. Idempotent commits via explicit `commit_message` per upload.
25. `promote.yml` uses `git push --force-with-lease=main:<MAIN_TIP>`; a diverged main refuses the push.

**Verification**
- `docker compose -p fed-pulse-fix320 run --rm backend pytest tests/unit/test_hf_registry_resolver.py tests/unit/test_embedding_cache_lazy_fetch.py tests/unit/test_reproduce_all_smoke.py -q` — 26 passed, 1 skipped.
- `ruff check app/models/registry.py app/data/embedding_cache.py app/training/loaders.py` — clean.
- `mypy --strict app/models/registry.py` — clean; `loaders.py` clean; `embedding_cache.py` 15 errors all pre-existing.
- `cd frontend && NEXT_PUBLIC_API_URL=https://test.example.com npm run build` — bundle bakes the URL (verified via grep into `.next/standalone/`).

Wiki §13 burn-down updated to enumerate the addressed findings.
